Vanya b/refactoring

VanyaBenovska opened the pull request.

Commit:
[=] js to ts files
[+] adding libraries and type libraries
[+] tsconfig
[+] setRoutes for route definitions
[+] startServer
[+] getZodiac endpoint
[+] interfaces IRestServer, IRestServerConfiguration / server.ts
[+] add text in file - mergeTextToFile / files.ts
[+] Adding enums
[+] validateEnvironmentVariable / helpers.ts
[+] constants / errors.ts
[+] add generated files

Commit:
[+] db Firebase, Cloud Firestore require("firebase-admin")
[+] daily text in sings' files
[+] function convertMapToObject for rest answer
[+] f to generate document name by date for DB document collections (for each sign separate documents' collection)
[+] f creates new document with the daily sign information to db
[+] f appends daily signs' information to the existing text in the document "AllSignsInfoInOneString" in db
[+] f creates document for all signs information if no such document in db
[=] f mergeTextToFile
[-] f saveToFile
[+] interface ISignsBGtexts
[=] add middleware processingRoutes
[+] main logic f for signs' daily info: GetDailyInfo, getDocumentNameByDate, processSignInfo, recordAbsent,  allSignsInformationDocumentAbsent, mergeTextToFile, etc.

Commit:
[=] DB changed Zodiac to ZodiacDeep
[+] Zodiac DB f (deprecated)
[+] Zodiacdeep DB business logic
[+] f to get Random number, Random short dateString, etc.
[+] logic for getting Lucky text from DB based on Random number
[+] SignData class
[=] Refactoring

Commit:
[-] Zodiac business logic (old DB Zodiac)
[-] f not used

Commit:
[=] By.xpath -> to  By.css finding web element
[+] Lucky rest request business logic (previous build)
[=] asynchronous functions
[=] business logic sending rest response for "lucky" request

Commit:
[+] business logic for automatic daily downloading all signs info
[+] winston lib and business logic for error handling
[-] functions for writing in files
[+] business logic to get info and operate from/with DB

Commit:
[+] css -> getAttribute("innerText")
[+] Jest library
[+] tests
[-] some repeating functions

Commit:
[+] timer code to get all signs data daily
[+] returning all signs' data for today if wrong query

jekovniki commented on files/bliznatsi.txt:
Delete all files inside of the 'files' folder (and the folder itself). This should be stored in the database.

jekovniki commented on __tests__/example.test.ts:
```diff
@@ -0,0 +1,54 @@
+type GetTestingRequest = {
```
This entire file is only example. You can delete it. It shouldn't be on production

jekovniki commented on __tests__/zodiac.test.ts:
```diff
@@ -0,0 +1,53 @@
+      .setChromeOptions(options.addArguments("--headless=new"))
+      .build();
+  });
+  test("should return sign info", async () => {
```
This test is testing only webdriver. You should test your own functionality and not the functionality of the selenium.

VanyaBenovska replied:
Yes, but we noticed some cases that information was not fetched, I really need this test

jekovniki commented on src/controllers/lucky.ts:
```diff
@@ -0,0 +1,30 @@
+import { getText } from "../helpers/arrayProcessing";
+import { getTodayAllSignRecordsFromDB } from "../dal/zodiacdeep";
+
+export async function getLuckyTextByRandomNumber(
```
The name of the function is **getLuckyTextByRandomNumber**, yet there is no random number passed. Something like **getLuckyZodiac** for name looks more appropriate;

Also this function can be put inside of **controllers/zodiac.ts** since it's related to zodiac

jekovniki commented on src/controllers/lucky.ts:
```diff
@@ -0,0 +1,30 @@
+  {
+    const luckyText = req?.query?.lucky;
+    let totalResult = "";
+    if (typeof luckyText !== "undefined" && luckyText === "lucky") {
```
I'd say remove the query. By naming I'm left with the impression that this method will return only lucky results, but it is returning the standard records from the database if there is no query.

I'd say remove the query and just keep only the logic on row 15.

VanyaBenovska replied:
done

jekovniki commented on src/controllers/lucky.ts:
```diff
@@ -0,0 +1,30 @@
+      if (resultTextArray) {
+        totalResult = getText(resultTextArray);
+      } else {
+        console.log("Not ok getting from db..");
```
This text is not saying much - you can replace it with the standard ERROR 404 Not found test and instead of logging it you can return it to the user.

VanyaBenovska replied:
done

jekovniki commented on src/controllers/lucky.ts:
```diff
@@ -0,0 +1,30 @@
+  req: Request,
+  res: Response
+): Promise<void> {
+  {
```
Put the code in try-catch

VanyaBenovska replied:
done

jekovniki commented on src/controllers/zodiac.ts:
```diff
@@ -0,0 +1,40 @@
+import { getText } from "../helpers/arrayProcessing";
+
+export async function getZodiac(req: Request, res: Response): Promise<void> {
+  {
```
Put the code in try-catch block

jekovniki commented on src/controllers/zodiac.ts:
```diff
@@ -0,0 +1,40 @@
+
+export async function getZodiac(req: Request, res: Response): Promise<void> {
+  {
+    const sign = req?.query?.sign;
```
here you can cast sign to string, so you don't have to call it everywhere else. Example:

const sign = req?.query?.sign as string;

jekovniki commented on src/controllers/zodiac.ts:
```diff
@@ -0,0 +1,40 @@
+    const sign = req?.query?.sign;
+    let totalResult = "";
+
+    if (typeof sign !== "undefined") {
```
Here you can reduce the if-else nesting by combining both cases and returning the response instead of assigning it.

jekovniki commented on src/controllers/zodiac.ts:
```diff
@@ -0,0 +1,40 @@
+    let totalResult = "";
+
+    if (typeof sign !== "undefined") {
+      if (Signs_ZodiacDirBG.includes(sign as string)) {
```
This can be combined with the upper if statement, so it becomes:

if (typeof sign !== 'undefined' && Signs_ZodiacDirBG.includes(sign))

jekovniki commented on src/controllers/zodiac.ts:
```diff
@@ -0,0 +1,40 @@
+        if (element) {
+          result.push(element);
+        }
+        totalResult = getText(result);
```
instead of assigning it, you can end the function here with a simple

```
res.send({ totalResult: getText(result) });
return
```

jekovniki commented on src/controllers/zodiac.ts:
```diff
@@ -0,0 +1,40 @@
+      } else if ((sign as string).toLowerCase() === "zodiac") {
+        const resultTextArray = await getTodayAllSignRecordsFromDB();
+        if (resultTextArray) {
+          totalResult = getText(resultTextArray);
```
instead of assigning it, you can end the function here with a simple

```
res.send({ totalResult: getText(result) });
return
```

jekovniki commented on src/controllers/zodiac.ts:
```diff
@@ -0,0 +1,40 @@
+    } else {
+      const resultTextArray = await getTodayAllSignRecordsFromDB();
+      if (resultTextArray) {
+        totalResult = getText(resultTextArray);
```
instead of assigning it, you can end the function here with a simple

```
res.send({ totalResult: getText(result) });
return
```

jekovniki commented on src/controllers/zodiac.ts:
```diff
@@ -0,0 +1,40 @@
+          totalResult = getText(resultTextArray);
+        }
+      }
+    } else {
```
Here if you have made the changes on the top if cases, you can skip the 'else' entirely.

jekovniki commented on src/controllers/zodiac.ts:
```diff
@@ -0,0 +1,40 @@
+          result.push(element);
+        }
+        totalResult = getText(result);
+      } else if ((sign as string).toLowerCase() === "zodiac") {
```
Here you can get out of this nested else if with

if (typeof sign !== 'undefined' && sign.toLowerCase() === "zodiac")

jekovniki commented on src/controllers/zodiac.ts:
```diff
@@ -0,0 +1,40 @@
+      const resultTextArray = await getTodayAllSignRecordsFromDB();
+      if (resultTextArray) {
+        totalResult = getText(resultTextArray);
+      }
```
NOTE:

I can see that you are doing - getText() after receiving each result. You can abstract it in a separate function OR if the idea of your function is to return the value from getText() all the time you can incorporate it in the other method.

jekovniki commented on src/dal/lucky.ts:
```diff
@@ -0,0 +1,29 @@
+import { getRandomNumberBetweenTwoNumbers } from "../randomizer";
+import { numbersSignsBG_zodiacDirBg } from "../models/signs";
+
+export async function getLuckyTextByRandomNumber(): Promise<string> {
```
Here the naming lies, you're not getting lucky text by random number, you're not passing any random number, so this function might have to be called **getLuckyText**?

jekovniki commented on src/dal/lucky.ts:
```diff
@@ -0,0 +1,29 @@
+  } catch (err) {
+    console.log(err);
+  }
+  return "";
```
Why are we returning empty string?

jekovniki commented on src/dal/lucky.ts:
```diff
@@ -0,0 +1,29 @@
+export async function getLuckyTextByRandomNumber(): Promise<string> {
+  const luckyNumber = getRandomNumberBetweenTwoNumbers(1, 12);
+  // Note! Sign names up to https://zodiac.dir.bg
+  const signObj = numbersSignsBG_zodiacDirBg.find(element => element.id === luckyNumber);
```
I see business logic here and I don't want to. This is DAL - Data Access Layer, which means that I want to see only a call to the database and return of result here. All of the business logic from here should be extracted into a separate function in services folder.

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+import { database } from "../libs/database";
```
What is zodiacdeep?

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+import { Signs_ZodiacDirBG } from "../models/signs";
+import { ISignRecord } from "../interfaces/signs";
+
+export async function GetArrayJSNatFromSignData(
```
use camelCase for naming. Also rethink the name of this function.

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+
+export async function GetArrayJSNatFromSignData(
+  sign: string
+): Promise<[{ sign: string; createdAt: string; text: string }]> {
```
Here instead of using this, you can use: ISignRecord[]

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+  } catch (err) {
+    console.log(err);
+  }
+  return [{ sign: sign, createdAt: "", text: "" }];
```
No, why are you returning this if it does not have any error?

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+  return [{ sign: sign, createdAt: "", text: "" }];
+}
+
+export async function GetLastElementJSNatFromSignData(
```
use camelCase for naming. Also rethink the name of this function. What is JSNat?

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+/*
+// Add today data object to the end of the data array 
+*/
+export async function AppendToArrayTodayElement(
```
This should not be in DAL. This is helper function :) . Also - use camelCase

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+  if (obj.createdAt === todayShortString) {
+    return false;
+  }
+  console.log("TODAY DATE IS NOT IN THE DATA!");
```
Remove this console.log

VanyaBenovska replied:
done

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+  return arrayData;
+}
+
+export async function isTodayRecordAbsent(
```
Instead of doing entire trip to the database - just pass as object the property to this function and then return the boolean. This way you can move this function into helpers folder

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+  return true;
+}
+
+export async function getTodayAllSignRecordsFromDB(): Promise<Array<ISignRecord> | void> {
```
There is a typo in the returning interface. Also for the naming - you don't need to specify that it is from the database. Also this function does not make a trip to the database so it's place should not be here

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+      signDataRef.set({ data: updatedArrayData });
+    }
+
+    // //******************** TODO: Refactoring
```
Remove this comment.

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+    if (!doc.exists) {
+      console.log("No such document!");
+    } else {
+      const currArrayData = await GetArrayJSNatFromSignData(sign);
```
You're doing both get and set operation in one function. Follow the SOLID principles (and more important the S of it).

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+ * @param tableId
+ * @param text
+ */
+export async function addSignDailyInfoIntoDB(
```
No need to specify "IntoDB"

jekovniki commented on src/helpers/files.ts:
```diff
@@ -0,0 +1,15 @@
+import fs from "fs";
```
This function is not used anywhere - you can delete this file.

VanyaBenovska replied:
done

jekovniki commented on src/helpers/id-generator.ts:
```diff
@@ -0,0 +1,7 @@
+export function getDateShortString(): string {
```
The function is great - a perfect example for helper function.

However the naming of the file threw me off. What is id-generator? Here we're not generating any id.

jekovniki commented on src/index.ts:
```diff
@@ -0,0 +1,55 @@
+import { handleErrors } from "./utils/errors";
+import { downloadLatestSignsData } from "./services/zodiacdeep";
+
+//// zodiac?sign=leo
```
Remove the comments

VanyaBenovska replied:
done

jekovniki commented on src/index.ts:
```diff
@@ -0,0 +1,55 @@
+  }
+})();
+
+export async function dailyDownloadLatestSignsData(): Promise<void> {
```
The place of this function is not in the main index. file. From what I see - this is business logic.

jekovniki commented on src/index.ts:
```diff
@@ -0,0 +1,55 @@
+  }
+}
+
+export async function dailyDownloadLatestSignsDataFORTESTS(): Promise<void> {
```
The place of this function is not in the main index. file. From what I see - this is some sort of test, and it should not be here (to be honest I believe it can be removed.

jekovniki commented on src/dal/lucky.ts:
```diff
@@ -0,0 +1,29 @@
+      return luckyText;
+    }
+  } catch (err) {
+    console.log(err);
```
You have handleErrors helper function in your project -  use it :) .

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+      return doc.get("data");
+    }
+  } catch (err) {
+    console.log(err);
```
You have handleErrors helper function in your project. Use it :) .

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+    }
+    return signResult;
+  } catch (err) {
+    console.log(err);
```
You have handleErrors helper function in your project. Use it :) .

jekovniki commented on src/dal/zodiacdeep.ts:
```diff
@@ -0,0 +1,126 @@
+    // });
+    // //********************
+  } catch (err) {
+    console.log(err);
```
You have handleErrors helper function in your project. Use it :) .

jekovniki commented on src/libs/rest.ts:
```diff
@@ -0,0 +1,9 @@
+
+export function startServer() : void {
+  app.listen(3000, () => {
+    console.log("Listening on port 3000 new");
```
You can use the logger that you have, instead of using console.log(). This is why you have the logger class in the first place.

Also what is 3000 new?

VanyaBenovska replied:
done

jekovniki commented on src/models/signData.ts:
```diff
@@ -0,0 +1,14 @@
+import { ISignRecord } from "../interfaces/signs";
+
+export class SignData implements ISignRecord{
```
I don't think you need this class

VanyaBenovska replied:
we desided to live it

jekovniki commented on src/models/signs.ts:
```diff
@@ -0,0 +1,194 @@
+export const zodiacDirBg_urlPart1 = `https://zodiac.dir.bg/sign/`;
```
Please refactor this -> I wan to see only constants that are used

jekovniki commented on src/models/xpaths.ts:
```diff
@@ -0,0 +1,17 @@
+// TODO: to map <url, xpath> + sign translated?
```
This is not used anywhere anymore - you can delete it.

jekovniki commented on src/models/signs.ts:
```diff
@@ -0,0 +1,194 @@
+export const zodiacDirBg_urlPart1 = `https://zodiac.dir.bg/sign/`;
+export const zodiacDirBg_urlPart2 = `/dneven-horoskop`;
```
Also the place for this code is not in models folder

jekovniki commented on src/randomizer.ts:
```diff
@@ -0,0 +1,18 @@
+/**
```
This file is perfect example for file with helpers. Move it to helpers folder and find more appropriate name for the file.

jekovniki commented on src/randomizer.ts:
```diff
@@ -0,0 +1,18 @@
+  min: number,
+  max: number
+): number {
+    max += 1;
```
maybe assign this to a meaningful constant at the moment.

jekovniki commented on src/routes/index.ts:
```diff
@@ -0,0 +1,22 @@
+      res.send("hello world");
+    });
+
+    // http://localhost:3000/zodiac?sign=leo
```
Remove the commented line.

VanyaBenovska replied:
done

jekovniki commented on src/routes/index.ts:
```diff
@@ -0,0 +1,22 @@
+    app.get("/lucky", loggingMiddleware, getLuckyTextByRandomNumber);
+    
+  } catch (error) {
+    console.log(error);
```
You have handleErrors that is designed exactly for these cases.

VanyaBenovska replied:
done

jekovniki commented on src/utils/constants/errors.ts:
```diff
@@ -0,0 +1,4 @@
+export const UNEXPECTED = "Unexpected error";
```
You're not using these constants anywhere - just remove the file

jekovniki commented on src/services/zodiacdeep.ts:
```diff
@@ -0,0 +1,109 @@
+promise.USE_PROMISE_MANAGER = false;
+
+// https://zodiac.dir.bg
+export const SignsBGtexts = new Map<string, string>([
```
You're not using this constant anywhere

VanyaBenovska replied:
done

jekovniki commented on src/services/zodiacdeep.ts:
```diff
@@ -0,0 +1,109 @@
+  } catch (err) {
+    console.log(err);
+  } finally {
+    console.log("QUIT DRIVER.");
```
remove console.log

VanyaBenovska replied:
done

jekovniki commented on src/services/zodiacdeep.ts:
```diff
@@ -0,0 +1,109 @@
+      );
+    }
+  } catch (err) {
+    console.log(err);
```
handleErrors

jekovniki commented on src/services/zodiacdeep.ts:
```diff
@@ -0,0 +1,109 @@
+  } catch (error) {
+    console.log(error);
+  }
+  return testResult;
```
Do we need to return testResult here?

jekovniki commented on src/services/zodiacdeep.ts:
```diff
@@ -0,0 +1,109 @@
+    await driver.get(url);
+    const dateToShortString = getDateShortString();
+    const isAbsent = await isTodayRecordAbsent(sign, dateToShortString);
+    console.log("Is data absent in DB: " + isAbsent);
```
Remove unnecessary console.logs and unnecessary comments

jekovniki commented on src/services/zodiacdeep.ts:
```diff
@@ -0,0 +1,109 @@
+      const content = await signText.getAttribute("innerText");
+
+      //"#content > div.article-body.horoscope p"))
+      if (signText) {
```
Try to replace the if elses with early exists of the function with **return** statement

VanyaBenovska replied:
done

Commit:
[-] not used functions, files, classes and comments
[=] code refactoring: move some files to proper directories
[+] reusing logger (handleErrors) in all places for error handling
